include/ginkgo/core/base/types.hpp

```diff
@@ -193,6 +196,179 @@ using default_precision = double;
 constexpr size_type byte_size = CHAR_BIT;
 
 
+namespace detail {
+
+
+/**
+ * mask gives the integer with Size activated bits in the end
+ *
+ * @tparam Size  the number of activated bits
+ * @tparam ValueType  the type of mask, which uses unsigned int as default
+ *
+ * @return the ValueType with Size activated bits in the end
+ */
+template <int Size, typename ValueType = unsigned int>
+constexpr std::enable_if_t<(Size < sizeof(ValueType) * 8), ValueType> mask()
+{
+    return (ValueType{1} << Size) - 1;
+}
+
+/**
+ * @copydoc mask()
+ *
+ * @note this is special case for the Size = the number of bits of ValueType
+ */
+template <int Size, typename ValueType = unsigned int>
+constexpr std::enable_if_t<Size == sizeof(ValueType) * 8, ValueType> mask()
+{
+    return ~ValueType{};
+}
+
+
+/**
+ * shift calculates the number of bits for shifting
+ *
+ * @tparam current_shift  the current position of shifting
+ * @tparam num_groups  the number of elements in array
+ *
+ * @return the number of shifting bits
+ *
+ * @note this is the last case of nested template
+ */
+template <int current_shift, int num_groups>
+constexpr std::enable_if_t<(num_groups == current_shift + 1), int> shift(
+    const std::array<unsigned char, num_groups> &bits)
+{
+    return 0;
+}
+
+/**
+ * @copydoc shift(const std::array<char, num_groups>)
+ *
+ * @note this is the usual case of nested template
+ */
+template <int current_shift, int num_groups>
+constexpr std::enable_if_t<(num_groups > current_shift + 1), int> shift(
+    const std::array<unsigned char, num_groups> &bits)
+{
+    return bits[current_shift + 1] +
+           shift<(current_shift + 1), num_groups>(bits);
+}
+
+
+}  // namespace detail
+
+
+// a shortcut type for ConfigSet
+using ConfigSetType = unsigned int;
+
+/**
+ * ConfigSet is a way to embed several information into one integer by given
+ * certain bits.
+ *
+ * The usage will be the following
+ * Set the method with bits Cfg = ConfigSet<b_0, b_1, ..., b_k>
+ * Encode the given infomation encoded = Cfg::encode(x_0, x_1, ..., x_k)
+ * Decode the specific position information x_t = Cfg::decode<t>(encoded)
+ * The encoded result will use 32 bits to record
+ * rrrrr0..01....1...k..k, which 1/2/.../k means the bits store the information
+ * for 1/2/.../k position and r is for rest of unused bits.
+ *
+ * Denote B_t = sum_(t+1)^(k) b_i and F(X) = Cfg::encode(x_0, ..., x_k)
```

Comment:
Can you use `$` to wrap the math portions, that should provide a better documentation style?